---
title: "Automatic Annotation Script for Skin and HF scRNA sequence data 01"
author: "Victor Ubels"
date: "November 2023"
output:
  html_notebook: 
    toc: yes
  html_document:
    keep_md: yes
    highlight: pygments
    df_print: paged
    toc: yes
    toc_float: true
editor_options: 
  chunk_output_type: inline
---

```{r}
### Parameters for batch processing of scRNA files
r_files <- list.files(pattern = "\\.R$", full.names = TRUE)

# Source each file
for (file in r_files) {
  source(file)
}

dir.create(paste0(output_directory_base, "/", project_name))

options(future.globals.maxSize = nThreads * 1000 * 1024^2)

#Set up work directory
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_knit$set(root.dir = output_directory)

load_libraries(libraries)

# Use the function
pymodules <- setup_python_environment()
pymagic <- pymodules$magic
```

Batch reading MTX files
```{r}
# Call the function with the correct directory path
read_all_mtx_and_create_seurat(directory = directory_mtx)
message("All files read")
```

Pre-processing data before doublet and ambient RNA removal
```{r}
# Retrieve all objects in the global environment that match the pattern "PATTERN"
seurat.objects_names <- pattern_files

# Get the actual Seurat objects based on the names
seurat.objects <- mget(seurat.objects_names, envir = .GlobalEnv)

# Apply the processing function to each Seurat object
processed.seurats <- lapply(seurat.objects, Process_Seurat)

# Now, 'processed_seurats' is a list of processed Seurat objects. 
# To replace the original Seurat objects in your global environment with the processed ones:
for (i in seq_along(processed.seurats)) {
  assign(seurat.objects_names[i], processed.seurats[[i]], envir = .GlobalEnv)
}

# At this point, your original Seurat objects in the global environment have been replaced with the processed ones.
rm(processed.seurats)
rm(seurat.objects)
```

Doublet removal
```{r}
dir.create(paste0(output_directory_base, project_name, "/Clustering_01/"))

# Get the actual Seurat objects based on the names
seurat.objects <- mget(seurat.objects_names, envir = .GlobalEnv)

# Apply the processing function to each Seurat object
processed.seurats <- Map(Process_Doublet_Removal, seurat_obj = seurat.objects, seurat_names = seurat.objects_names)

# Now, 'processed_seurats' is a list of processed Seurat objects. 
# To replace the original Seurat objects in your global environment with the processed ones:
for (i in seq_along(processed.seurats)) {
  assign(seurat.objects_names[i], processed.seurats[[i]], envir = .GlobalEnv)
}


rm(processed.seurats)
rm(seurat.objects)
```

Ambient RNA removal
```{r}
seurat.objects <- mget(seurat.objects_names, envir = .GlobalEnv)

seurat.objects <- lapply(seurat.objects, runDecontX)
seurat.objects <- lapply(seurat.objects, Remove_lowRNA)

for (i in seq_along(seurat.objects)) {
  assign(seurat.objects_names[i], seurat.objects[[i]], envir = .GlobalEnv)
}

rm(seurat.objects)
```

Applies diseaseStatus to each object (Based on filename, groups objects based on experiment)
```{r}
seurat.objects <- mget(seurat.objects_names, envir = .GlobalEnv)

for (i in 1:length(seurat.objects)) {
  seurat.objects[[i]]@meta.data$orig.ident <- seurat.objects_names[i]
}

get_disease_status <- function(name) {
  # Extract the part after GSM number and underscore
  match <- regexpr("GSM\\d+_([A-Z_]+)", name)
  if (match != -1) {
    extracted <- regmatches(name, match)[[1]]
    # Extract everything after the GSM number and underscore, up to any digits
    status <- sub("GSM\\d+_([A-Z_]+).*", "\\1", extracted)
    return(status)
  } else {
    return("Unknown")
  }
}

for (i in 1:length(seurat.objects)) {
  # Assign orig.ident
  seurat.objects[[i]]@meta.data$orig.ident <- names(seurat.objects)[i]
  
  # Determine and assign disease status
  disease_status <- get_disease_status(names(seurat.objects)[i])
  seurat.objects[[i]]@meta.data$diseaseStatus <- disease_status
}

list2env(setNames(seurat.objects, names(seurat.objects)), envir = .GlobalEnv)

seurat_obj_merged <- merge(x=seurat.objects[[1]], y=seurat.objects[2:length(seurat.objects)], project=project_name)

rm(list = seurat.objects_names)
rm(seurat.objects)
```

Predetermines genes to ignore for clustering. Ignored genes are mitotic cycle genes, X/Y chromosome genes and mitochondrial genes.
```{r}
rawCounts <- GetAssayData(object=seurat_obj_merged, slot="counts")

# Identify genes we want to blacklist during clustering

mt.genes <- grep(pattern = "^MT-", x = rownames(rawCounts), value = TRUE)
RPS.genes <- grep(pattern = "^RPS", x = rownames(rawCounts), value = TRUE)
RPL.genes <- grep(pattern = "^RPL", x = rownames(rawCounts), value = TRUE)
  
# X/Y chromosome genes:
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
geneGR <- GenomicFeatures::genes(txdb)
sexGenesGR <- geneGR[seqnames(geneGR) %in% c("chrY", "chrX")]
matchedGeneSymbols <- select(org.Hs.eg.db,
                        keys = sexGenesGR$gene_id,
                        columns = c("ENTREZID", "SYMBOL"),
                        keytype = "ENTREZID")
sexChr.genes <- matchedGeneSymbols$SYMBOL


# Genes to ignore (just for clustering purposes)
blacklist.genes <- c(
    mt.genes,
    sexChr.genes,
    s.genes,
    g2m.genes,
    RPS.genes,
    RPL.genes
)

seurat_obj_merged <- CellCycleScoring(seurat_obj_merged, s.features=s.genes, g2m.features=g2m.genes, set.ident=FALSE)
```

Returns filtered barcode sets for each sample after quality control

```{r}
sample_ID <- unique(seurat_obj_merged@meta.data$orig.ident)

for (i in seq_along(sample_ID)) {

  sample <- sample_ID[i]
  sub_obj <- subset(seurat_obj_merged, subset = orig.ident == sample)
  filtered_barcodes <- sub_obj@assays$RNA@data@Dimnames[[2]]
  filtered_barcodes <- sub("-1_[0-999]+$", "-1", filtered_barcodes)
  filtered_barcodes <- as.data.frame(filtered_barcodes)
  write_tsv(filtered_barcodes, file = paste0(output_directory, "filtered_", sample ,"_barcodes.tsv"), col_names = FALSE)

}

rm(sub_obj)
```

Log normalization and initial broad cluster allocation

```{r}
lsiOut <- list()
clusters <- NULL

# Depth normalize to 10,000, add pseudo count, and then log2 transform
log2CP10k <- sparseLogX(rawCounts, logtype="log2", scale=TRUE, scaleFactor=10^4)
# Store the log2CP10k
seurat_obj_merged <- SetAssayData(object=seurat_obj_merged, slot="data", new.data=log2CP10k)

# UMAP
umapNeighbors <- 50
umapMinDist <- 0.5
umapDistMetric <- "cosine"


message("Running iterative LSI...")
set.seed(1)

#Initial Cluster allocation

for(i in seq_along(resolution)){
    # If first round, compute variable genes on raw data first
    if(i == 1){
        message(sprintf("Identifying top %s variable genes among all cells...", nVarGenes))
        varGenes <- getVarGenes(log2CP10k, nvar=nVarGenes, blacklist=blacklist.genes)
    }else{
        # For remaining rounds, calculate variable genes using previous clusters
        clusterMat <- edgeR::cpm(groupSums(rawCounts, clusters, sparse=TRUE), log=TRUE, prior.count=3)
        message(sprintf("Identifying top %s variable genes from round %s LSI...", nVarGenes, i-1))
        varGenes <- getVarGenes(clusterMat, nvar=nVarGenes, blacklist=blacklist.genes)
    }
    # Now run LSI and find clusters
    LSIi <- runLSI(rawCounts[varGenes,], nComponents = max(nPCs), binarize = FALSE)

    # 'Harmonize' SVD PCs, if indicated
    if(i %in% harmonize){
      message(sprintf("Harmonizing LSI SVD PCs for round %s...", i))
      harmonized_pcs <- HarmonyMatrix(
        data_mat  = LSIi$matSVD,
        meta_data = seurat_obj_merged@meta.data,
        vars_use  = covariates, # Covariates to 'harmonize'
        do_pca    = FALSE
        )
      LSIi$matSVD <- harmonized_pcs
    }

    reducName <- paste0("LSI_iter",i)
    seurat_obj_merged[[reducName]] <- CreateDimReducObject(embeddings = LSIi$matSVD, key = sprintf("LSI%s_", i), assay = "RNA")
    seurat_obj_merged <- FindNeighbors(object = seurat_obj_merged, reduction = reducName, dims = nPCs, force.recalc = TRUE)
    message(sprintf("Clustering with resolution %s...", resolution[i]))
    seurat_obj_merged <- FindClusters(object = seurat_obj_merged, resolution = resolution[i], algorithm = 4, method = "igraph")
    clusters <- Idents(seurat_obj_merged)
    #Store information
    lsiOut[[reducName]] <- list(
        lsiMat = LSIi$matSVD,
        svd = LSIi$svd,
        varFeatures = varGenes, 
        clusters = clusters
    )
}

# Store cluster information in metadata
seurat_obj_merged$Clusters <- Idents(seurat_obj_merged)
```

```{r}
message("Calculating UMAP...")

set.seed(1)

seurat_obj_merged <- RunUMAP(
  seurat_obj_merged,
  reduction = paste0("LSI_iter",length(resolution)), # Use final LSI iteration 
  dims = nPCs,
  n.neighbors = umapNeighbors,
  min.dist = umapMinDist,
  metric = umapDistMetric,
  verbose = TRUE
)

#Generates a Preclustering directory in the output directory
dir.create(paste0(output_directory, "/Clustering_01"))

# Save clustered object here:
message("Saving Seurat object...")
saveRDS(seurat_obj_merged, file = paste0(output_directory, "Clustering_01/Cluster_01_Seurat_Object.rds"))

# Save iterativeLSI info
message("Saving LSI information...")
saveRDS(lsiOut, file = paste0(output_directory, "/Clustering_01/lsiOut_Cluster_01_Seurat_Object.rds.rds"))

message("Done.")

DimPlot(seurat_obj_merged, label = TRUE, split.by = 'diseaseStatus')
ggsave(filename = paste0(output_directory, "Clustering_01/Split_By_Disease.pdf"), plot = last_plot(), width = 20)
```

```{r}
markers <- FindAllMarkers(seurat_obj_merged,
                          logfc.threshold = 1.5,
                          min.pct = 0.25,
                          only.pos = TRUE,
                          test.use = "roc"
                        )

unique_markers <- markers %>%
  group_by(gene) %>%
  mutate(n_clusters = n(),
         uniqueness_score = avg_log2FC / n_clusters) %>%
  ungroup() %>%
  arrange(cluster, desc(uniqueness_score))

top_unique_markers <- unique_markers %>%
  group_by(cluster) %>%
  top_n(n = 20, wt = uniqueness_score)

single_cluster_markers <- unique_markers %>%
  filter(n_clusters == 1) %>%
  group_by(cluster) %>%
  top_n(n = 20, wt = avg_log2FC)

clusters_without_unique_genes <- setdiff(unique(markers$cluster), 
                                         unique(single_cluster_markers$cluster))

for (cluster in clusters_without_unique_genes) {
  cluster_markers <- unique_markers %>%
    filter(cluster == !!cluster) %>%
    top_n(n = 5, wt = uniqueness_score)
  
  single_cluster_markers <- bind_rows(single_cluster_markers, cluster_markers)
}

output_filename_markers <- paste0(output_directory, "/Clustering_01/Markers_PreClustering", dataset, ".xlsx")
write_xlsx(MarkerTibble, output_filename_markers)

winDialog("ok", "Script finished running")
```
